ssl/ssl_lib.c

```diff
@@ -3294,10 +3312,24 @@ int SSL_CTX_set_cipher_list(SSL_CTX *ctx, const char *str)
      */
     if (sk == NULL)
         return 0;
-    else if (cipher_list_tls12_num(sk) == 0) {
+    /*
+     * check whether cipher and protocl match
+     */
+    if (cipher_list_tls12_num(sk) == 0 && cipher_list_tls3_num(sk) == 0) {
```

Comment:
It's sensible to cache the returns from these two calls.  It's potentially a moderate amount of computation to count them.